drivers/disk/CMakeLists.txt

```diff
@@ -15,4 +17,91 @@ zephyr_library_sources_ifdef(CONFIG_MMC_SUBSYS mmc_subsys.c)
 
 add_subdirectory_ifdef(CONFIG_NVME nvme)
 
+set(gen_inc_dir ${ZEPHYR_BINARY_DIR}/misc/generated/ramdisk)
+set(generated_file ${gen_inc_dir}/ramdisk_diskimages.inc)
+
+zephyr_include_directories(${gen_inc_dir})
+
+# Scan RAM disk device nodes in the DTS to identify the ones that are enabled and
+# specify disk image paths.
+dt_comp_path(ramdisk_devs COMPATIBLE "zephyr,ram-disk")
+foreach(ramdisk_dev ${ramdisk_devs})
+
+  dt_node_has_status(dt_status_okay PATH ${ramdisk_dev} STATUS "okay")
+  dt_prop(dt_diskimage_path PATH ${ramdisk_dev} PROPERTY "zephyr,disk-image-path")
+  if(dt_status_okay AND DEFINED dt_diskimage_path)
+
+    list(APPEND ramdisk_dev_paths ${ramdisk_dev})
+
+    dt_prop(disk_sector_size PATH ${ramdisk_dev} PROPERTY "sector-size")
+    list(APPEND ramdisk_sector_sizes ${disk_sector_size})
+
+    cmake_path(IS_ABSOLUTE dt_diskimage_path is_diskimg_path_absolute)
+    if(${is_diskimg_path_absolute})
+
+      list(APPEND ramdisk_diskimg_paths ${dt_diskimage_path})
+
+    else()
+
+      # Relative paths are assumed to be relative to Zephyr build directory.
+      # Adjust the path accordingly.
+      list(APPEND ramdisk_diskimg_paths ${ZEPHYR_BINARY_DIR}/../${dt_diskimage_path})
+
+    endif()
+
+    # If GNU extensions are not available for use, disk image binary data needs to be converted into
+    # hex characters form so that it can be included as array data.
+    if(NOT CONFIG_TOOLCHAIN_SUPPORTS_GNU_EXTENSIONS)
+
+      string(TOLOWER ${ramdisk_dev} ramdisk_dev)
+      string(REPLACE "/" "_S_" _ramdisk_dev ${ramdisk_dev})
+      string(REGEX REPLACE "[-,.@/+]" "_" ramdisk_dev_path_id ${_ramdisk_dev})
+      string(PREPEND ramdisk_dev_path_id "N")
+      list(APPEND ramdisk_diskimg_data_incs ${gen_inc_dir}/${ramdisk_dev_path_id}.inc)
+
+    endif()
+
+  endif()
+
+endforeach()
+
+# If any RAM disk devices are found in the DTS, that use disk images as backing store,
+# add appropriate commands to include the disk images binary data in the build.
+if (DEFINED ramdisk_diskimg_paths)
+
+  add_custom_command(
```

Comment:
Is the reason for this custom command because we need this generation step to run at the end of the build? IE once all the include files have been generated for binaries? I'm not clear why we need to add a separate script here instead of just calling a CMake function or macro here to write the file.

Reply:
> Is the reason for this custom command because we need this generation step to run at the end of the build? IE once all the include files have been generated for binaries?

It is for build dependency management. If we just create the file (`ramdisk_diskimages.inc`) in-place here, the file will be generated fine in a clean build, or right after you have edited (or touched) this CMakeLists.txt file, but subsequently, if you only update the disk image file(s), `ramdisk_diskimages.inc` file will not be regenerated - so the built image will still be using the old stale disk image(s) content.

> I'm not clear why we need to add a separate script here instead of just calling a CMake function or macro here to write the file.

A separate script could have been avoided, if there were some way of specifying a CMake function as command of a custom target but I couldn't find any way. The closest option was to put the code in a separate file and invoke it as a CMake script, as the command of the custom build target.